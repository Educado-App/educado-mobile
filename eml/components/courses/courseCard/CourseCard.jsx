import { View, Pressable } from 'react-native';
import React, { useState } from 'react';
import { useNavigation } from '@react-navigation/native';
import Text from '../../../components/general/Text';
import { MaterialCommunityIcons } from '@expo/vector-icons';
import CustomProgressBar from '../../exercise/Progressbar';
import tailwindConfig from '../../../tailwind.config';
import { determineIcon, determineCategory, formatHours, checkProgressCourse} from '../../../services/utilityFunctions';
import DownloadCourseButton from './DownloadCourseButton';
import PropTypes from 'prop-types';
import { checkCourseStoredLocally } from '../../../services/StorageService';

/**
 * CourseCard component displays a card for a course with its details
 * @param {Object} props - Component props
 * @param {Object} props.course - Course object containing course details
 * @returns {JSX.Element} - Rendered component
 */
export default function CourseCard({ course, isOnline}) {
	const [downloaded, setDownloaded] = useState(false);
	const navigation = useNavigation();
	const [studentProgress, setStudentProgress] = useState(0);


	const checkDownload = async () => {
		setDownloaded(await checkCourseStoredLocally(course.courseId));
	};
	checkDownload();

	const checkProgress = async () => {
		const progress = await checkProgressCourse(course.courseId);
		setStudentProgress(progress);
	}; checkProgress();

	const enabledUI = 'bg-projectWhite m-[3%] rounded-lg shadow-sm shadow-opacity-[0.3] elevation-[8] mx-[5%] p-[5%]';
	const disabledUI = 'opacity-50 bg-projectWhite m-[3%] rounded-lg shadow-sm shadow-opacity-[0.3] elevation-[8] mx-[5%] p-[5%]';

	const layout = downloaded || isOnline ? enabledUI : disabledUI;

	let isDisabled = layout === disabledUI;

	return (
		<Pressable testID="courseCard"
			className={layout}
			onPress={() => { layout === enabledUI ?
				navigation.navigate('Section', {
					course: course,
				}) : null;
			}}
		>
			<View>
				<View className="flex-row items-start justify-between px-[1%] py-[1%]">
					<Text className="text-[18px] text-projectBlack flex-1 self-center font-montserrat-semi-bold">
						{course.title ? course.title : 'Título do curso'}
					</Text>
					<View className="flex-2 pr-6">
						<DownloadCourseButton course={course} disabled={isDisabled}/>
					</View>
				</View>
				<View className="h-[1] bg-disable m-[2%]" />
				<View className="flex-row flex-wrap items-center justify-start">
					<View className="flex-row items-center">
						<MaterialCommunityIcons size={18} name={determineIcon(course.category)} color={'gray'}></MaterialCommunityIcons>
						<Text className="mx-[2.5%] my-[3%]">{determineCategory(course.category)}</Text>
					</View>
					<View className="flex-row items-center">
						<MaterialCommunityIcons size={18} name="clock" color={'gray'}></MaterialCommunityIcons>
						<Text className="mx-[2.5%] my-[3%]">{course.estimatedHours ? formatHours(course.estimatedHours) : 'duração'}</Text>
					</View>
				</View>
				<View className="flex-row items-center">
					<CustomProgressBar width={56} progress={studentProgress} height={1} />
					<Pressable className="z-[1]"
						onPress={() => {layout === enabledUI ?
							navigation.navigate('Section', {
								course: course,
							}) : null;
						}}
					>
						<MaterialCommunityIcons size={28} name="play-circle" color={tailwindConfig.theme.colors.primary}></MaterialCommunityIcons>
					</Pressable>
				</View>
			</View>
		</Pressable>
	);
}

CourseCard.propTypes = {
	course: PropTypes.oneOfType([
		PropTypes.object,
		PropTypes.array,
	]),
	isOnline: PropTypes.bool
};

