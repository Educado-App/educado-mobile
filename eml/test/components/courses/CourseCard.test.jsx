import React from "react";
import renderer from "react-test-renderer";
import CourseCard from "../../../components/courses/courseCard/CourseCard";
import AsyncStorage from "@react-native-async-storage/async-storage";

let navigated = false;

jest.mock('@react-navigation/native', () => ({
    useNavigation: () => ({
      navigate: jest.fn (() => {navigated = true}),
    }),
  }));
  
    let courseCard;

    const courses = [
        {
          id: 1,
          title: "Course 1",
          category: "Category 1",
          duration: "Duration for course 1",
          image: "",
        },
        {
          id: 2,
          title: "Course 2",
          category: "Category 2",
          duration: "Duration for course 2",
          image: "",
        },
      ];

    beforeEach(() => {
        navigated = false;
        AsyncStorage.clear();
        courseCard = renderer.create(<CourseCard course={courses} />);
    });

    afterAll(() => {
        jest.resetModules();
        jest.restoreAllMocks();
    });

    test("renders CourseCard correctly", () => {
        expect(courseCard.toJSON()).toMatchSnapshot();
    });

    test('Navigate to section when pressing the course card', () => { 
        const button = courseCard.root.findByProps({testID: 'courseCard'});
        button.props.onPress();
        expect(navigated).toBe(true);
    });

