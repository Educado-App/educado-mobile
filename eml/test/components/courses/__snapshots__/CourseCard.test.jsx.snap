// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CourseCard renders CourseCard correctly 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "backgroundColor": "#FFFFFF",
      },
      {
        "marginBottom": "3%",
        "marginLeft": "3%",
        "marginRight": "3%",
        "marginTop": "3%",
      },
      {
        "borderBottomLeftRadius": 8,
        "borderBottomRightRadius": 8,
        "borderTopLeftRadius": 8,
        "borderTopRightRadius": 8,
      },
      {
        "shadowColor": "rgba(0, 0, 0, 0.1)",
        "shadowOffset": {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 1,
        "shadowRadius": 2,
      },
      {
        "elevation": 8,
      },
      {
        "marginLeft": "5%",
        "marginRight": "5%",
      },
      {
        "paddingBottom": "5%",
        "paddingLeft": "5%",
        "paddingRight": "5%",
        "paddingTop": "5%",
      },
    ]
  }
  testID="courseCard"
>
  <View>
    <View
      style={
        [
          {
            "flexDirection": "row",
          },
          {
            "alignItems": "flex-start",
          },
          {
            "justifyContent": "space-between",
          },
          {
            "paddingLeft": "1%",
            "paddingRight": "1%",
          },
          {
            "paddingBottom": "1%",
            "paddingTop": "1%",
          },
        ]
      }
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          [
            {
              "color": undefined,
              "fontSize": 28,
            },
            undefined,
            {
              "fontFamily": "material-community",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        󰑴
         
      </Text>
      <Text
        style={
          [
            {
              "fontFamily": "Montserrat-Regular",
            },
            [
              {
                "fontSize": 18,
              },
              {
                "flexBasis": "0%",
                "flexGrow": 1,
                "flexShrink": 1,
              },
              {
                "alignSelf": "center",
              },
            ],
          ]
        }
      >
        Course Title
      </Text>
    </View>
    <View
      style={
        [
          {
            "height": 1,
          },
          {
            "backgroundColor": "#DDD",
          },
          {
            "marginBottom": "2%",
            "marginLeft": "2%",
            "marginRight": "2%",
            "marginTop": "2%",
          },
        ]
      }
    />
    <View
      style={
        [
          {
            "flexDirection": "row",
          },
          {
            "alignItems": "center",
          },
          {
            "justifyContent": "flex-start",
          },
          {
            "overflow": "hidden",
          },
        ]
      }
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          [
            {
              "color": "gray",
              "fontSize": 18,
            },
            undefined,
            {
              "fontFamily": "material-community",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        󰑴
      </Text>
      <Text
        style={
          [
            {
              "fontFamily": "Montserrat-Regular",
            },
            [
              {
                "marginLeft": "2.5%",
                "marginRight": "2.5%",
              },
              {
                "marginBottom": "3%",
                "marginTop": "3%",
              },
            ],
          ]
        }
      >
        category
      </Text>
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          [
            {
              "color": "gray",
              "fontSize": 18,
            },
            undefined,
            {
              "fontFamily": "material-community",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        󰥔
      </Text>
      <Text
        style={
          [
            {
              "fontFamily": "Montserrat-Regular",
            },
            [
              {
                "marginLeft": "2.5%",
                "marginRight": "2.5%",
              },
              {
                "marginBottom": "3%",
                "marginTop": "3%",
              },
            ],
          ]
        }
      >
        duration
      </Text>
    </View>
    <View
      style={
        [
          {
            "flexDirection": "row",
          },
          {
            "alignItems": "center",
          },
        ]
      }
    >
      <View
        style={
          [
            {
              "flexDirection": "row",
            },
            {
              "alignItems": "center",
            },
            {
              "justifyContent": "space-around",
            },
          ]
        }
      >
        <View
          animated={true}
          animationConfig={
            {
              "bounciness": 0,
            }
          }
          animationType="spring"
          indeterminate={false}
          indeterminateAnimationDuration={1000}
          onLayout={[Function]}
          progress={0.5}
          style={
            [
              {
                "backgroundColor": "rgba(228, 242, 245, 1)",
                "borderColor": "rgba(94, 204, 233, 1)",
                "borderRadius": 8,
                "borderWidth": 0,
                "overflow": "hidden",
                "width": 450,
              },
              undefined,
            ]
          }
          useNativeDriver={false}
        >
          <View
            collapsable={false}
            style={
              {
                "backgroundColor": "rgba(94, 204, 233, 1)",
                "height": 13.34,
                "transform": [
                  {
                    "translateX": 0,
                  },
                  {
                    "translateX": -112.5,
                  },
                  {
                    "scaleX": 0.50005,
                  },
                ],
              }
            }
          />
        </View>
        <Text
          style={
            [
              {
                "paddingLeft": 20,
                "paddingRight": 20,
              },
              {
                "textAlign": "center",
              },
              {
                "fontFamily": "Montserrat-Bold",
              },
              {
                "fontSize": 12,
              },
              {
                "color": "#383838",
              },
            ]
          }
        >
          50
          %
        </Text>
      </View>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "zIndex": 1,
            },
          ]
        }
      >
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            [
              {
                "color": "#5ECCE9",
                "fontSize": 28,
              },
              undefined,
              {
                "fontFamily": "material-community",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              {},
            ]
          }
        >
          󰐌
        </Text>
      </View>
    </View>
  </View>
</View>
`;
